File: example/src/main/java/eu/darken/mvpbakery/example/screens/MainPresenter.java
package eu.darken.mvpbakery.example.screens;

import android.os.Bundle;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import eu.darken.mvpbakery.base.Presenter;
import eu.darken.mvpbakery.base.StateListener;
import eu.darken.mvpbakery.example.screens.counting.CountingFragment;
import eu.darken.mvpbakery.example.screens.text.TextFragment;
import eu.darken.mvpbakery.injection.ComponentPresenter;


@MainComponent.Scope
public class MainPresenter extends ComponentPresenter<MainPresenter.View, MainComponent> implements StateListener {

    private final Class<? extends Fragment> startingFragment;
    private boolean doInit = true;
    private int bindCounter = 0;
    private int fragmentNo = 1;

    @Inject
    MainPresenter(Class<? extends Fragment> startingFragment) {
        this.startingFragment = startingFragment;
    }

    @Override
    public void onRestoreState(@Nullable Bundle savedInstanceState) {
        if (savedInstanceState != null) doInit = savedInstanceState.getBoolean("doInit");
        if (savedInstanceState != null) bindCounter = savedInstanceState.getInt("counter");
        if (savedInstanceState != null) fragmentNo = savedInstanceState.getInt("fragmentNo");
    }

    @Override
    public void onSaveState(@NonNull Bundle outState) {
        outState.putBoolean("doInit", doInit);
        outState.putInt("counter", bindCounter);
        outState.putInt("fragmentNo", fragmentNo);
    }

    @Override
    public void onBindChange(@Nullable View view) {
        super.onBindChange(view);
        onView(v -> v.showBinderCounter(++bindCounter));
        onView(v -> {
            if (doInit) {
                doInit = false;
                v.addFragment(startingFragment);
            } else {
                int oldFrags = fragmentNo;
                fragmentNo = 0;
                for (int i = 0; i < oldFrags; i++) {
                    addNewFragment();
                }
            }
        });
    }

    private void addFragment(int pos) {
        switch (pos % 2) {
            case 0:
                onView(v -> v.addFragment(CountingFragment.class));
                break;
            case 1:
                onView(v -> v.addFragment(TextFragment.class));
                break;
        }
    }

    public void addNewFragment() {
        addFragment(fragmentNo++);
    }

    public interface View extends Presenter.View {
        void showBinderCounter(int count);

        void addFragment(Class<? extends Fragment> fragmentClass);
    }
}

File: example/src/main/java/eu/darken/mvpbakery/example/screens/MainPresenter.kt
package eu.darken.mvpbakery.example.screens

import android.os.Bundle
import androidx.fragment.app.Fragment
import eu.darken.mvpbakery.base.Presenter
import eu.darken.mvpbakery.base.StateListener
import eu.darken.mvpbakery.example.screens.counting.CountingFragment
import eu.darken.mvpbakery.example.screens.text.TextFragment
import eu.darken.mvpbakery.injection.ComponentPresenter
import javax.inject.Inject


@MainComponent.Scope
class MainPresenter @Inject
internal constructor(private val startingFragment: Class<out Fragment>) : ComponentPresenter<MainPresenter.View, MainComponent>(), StateListener {
    private var doInit = true
    private var bindCounter = 0
    private var fragmentNo = 1

    override fun onRestoreState(inState: Bundle?) {
        if (inState != null) doInit = inState.getBoolean("doInit")
        if (inState != null) bindCounter = inState.getInt("counter")
        if (inState != null) fragmentNo = inState.getInt("fragmentNo")
    }

    override fun onSaveState(outState: Bundle) {
        outState.putBoolean("doInit", doInit)
        outState.putInt("counter", bindCounter)
        outState.putInt("fragmentNo", fragmentNo)
    }

    override fun onBindChange(view: View?) {
        super.onBindChange(view)
        withView { v -> v.showBinderCounter(++bindCounter) }
        withView { v ->
            if (doInit) {
                doInit = false
                v.addFragment(startingFragment)
            } else {
                val oldFrags = fragmentNo
                fragmentNo = 0
                for (i in 0 until oldFrags) {
                    addNewFragment()
                }
            }
        }
    }

    private fun addFragment(pos: Int) {
        when (pos % 2) {
            0 -> withView { v -> v.addFragment(CountingFragment::class.java) }
            1 -> withView { v -> v.addFragment(TextFragment::class.java) }
        }
    }

    fun addNewFragment() {
        addFragment(fragmentNo++)
    }

    interface View : Presenter.View {
        fun showBinderCounter(count: Int)

        fun addFragment(fragmentClass: Class<out Fragment>)
    }
}
